src/game.rs

```diff
@@ -45,31 +98,61 @@ impl Game {
     }
 
     pub fn get_piece(&self, position: &Position) -> Option<Piece> {
-        self.state.get_piece(position)
+        self.history.get_current_state().get_piece(position)
     }
 
     pub fn get_board_state(&self) -> &BoardState {
-        &self.state
+        self.history.get_current_state()
     }
 
-    pub fn move_piece(&mut self, from: Position, to: Position) -> ChessResult {
-        if let Some(piece) = self.state.get_piece(&from) {
-            let mv = Move::new(from, to);
+    pub fn go_back_a_turn(&mut self) {
+        self.status = GameStatus::Replay;
+        self.history.previous_state()
+    }
 
-            self.is_move_valid(&mv)?;
-            self.state.move_piece(&mv);
-            self.add_moves();
-            self.remove_self_checks();
-            self.update_status();
+    pub fn go_forward_a_turn(&mut self) {
+        self.status = GameStatus::Replay;
+        self.history.next_state()
+    }
 
-            println!("{} : {}", piece, mv);
+    pub fn go_to_beginning(&mut self) {
+        self.status = GameStatus::Replay;
+        self.history.initial_state()
+    }
+
+    pub fn resume(&mut self) {
+        self.status = GameStatus::Ongoing;
+        self.history.resume()
+    }
+
+    pub fn move_piece(&mut self, from: Position, to: Position) -> ChessResult {
+        if self.status == GameStatus::Ongoing {
+            if let Some(piece) = self.history.get_current_state().get_piece(&from) {
+                let mv = Move::new(from, to);
+    
+                self.is_move_valid(&mv)?;
+                self.history.get_current_state_mut().move_piece(&mv);
```

Comment:
I think we can get the state immutably and clone it, and then push the cloned state on after making the move.